Review the SLAlpha4 Release Note (#2093)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-alpha4/swan-lake-alpha4-release-note.md

<em>This is the fourth Alpha release in a series of planned Alpha and Beta releases leading up to the Ballerina Swan Lake GA release.</em> 

> It introduces the new language features planned for the Swan Lake GA release and includes improvements and bug fixes done to the compiler, runtime, standard library, and developer tooling after the Swan Lake Alpha3 release.

Reviewer: this has to be alpha 4 right?

> It introduces the new language features planned for the Swan Lake GA release and includes improvements and bug fixes done to the compiler, runtime, standard library, and developer tooling after the Swan Lake Alpha3 release.

Reviewer: ```suggestion
It introduces the new language features planned for the Swan Lake GA release and includes improvements and bug fixes done to the compiler, runtime, standard library, and developer tooling after the Swan Lake Alpha 4 release.
```

Author: Nope, this means that we are doing the improvements on top of the SLAlpha3 release, which was the previous release.

Reviewer: Sorry my bad :)

<em>This is the fourth Alpha release in a series of planned Alpha and Beta releases leading up to the Ballerina Swan Lake GA release.</em> 

It introduces the new language features planned for the Swan Lake GA release and includes improvements and bug fixes done to the compiler, runtime, standard library, and developer tooling after the Swan Lake Alpha3 release.

- [Updating Ballerina](#updating-ballerina)
- [Installing Ballerina](#installing-ballerina)
- [Language Updates](#language-updates)
- [Runtime Updates](#runtime-updates)
- [Standard Library Updates](#standard-library-updates)
- [Developer Tools Updates](#developer-tools-updates)

> - [Security Updates](#security-updates)

Reviewer 2: ```suggestion
```

Author: Fixed.

#### Bug Fixes

To view bug fixes, see the [GitHub milestone for Swan Lake Alpha4](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FCompilerFE).

> ### Breaking Changes

Reviewer 3: Why is breaking changes here? The ToC indicates that it's at the end.

Author: That was updated as per https://github.com/ballerina-platform/ballerina-dev-website/pull/2093#discussion_r606619298. The TOC is also updated to no longer have the Breaking Changes as a separate topic.

Reviewer 4: To me, this doesn't seem to be correct. It is a level 3 heading similar to `Language updates`, `Runtime updates ` etc. So Is this a sub-topic under `Language updates`? In that case, it should be a level 4 heading.

Author: Updated.

> #### Breaking Changes

Reviewer 4: Shall we have this before the `Bug Fixes` section? Just to be align with other topics, so that `Bug Fixes` is the last section in each sub topic

Author: Updated.

#### Bug Fixes

To view bug fixes, see the [GitHub milestone for Swan Lake Alpha4](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FCompilerFE).

### Runtime Updates

#### New Features

> ##### Providing Values for Configurable Variables with Command Line Arguments

Reviewer 5: Shouldn't this be 

```suggestion
##### Providing Values for Configurable Variables via Command Line Arguments
```

Example usages:

```ballerina
configurable int port = ?;
```

If the configurable variable is defined in the default module or if we are using a single Ballerina file:

```bash
bal run -- -Cport=9090
bal run program.bal -- -Cport=9090
java -jar executable.jar -Cport=9090 
```

If the configurable variable is defined in a different module of the same organization:

```bash
bal run -- -Cmodule-name.port=9090
java -jar executable.jar -Cmodule-name.port=9090 
```

If the configurable variable is defined in a module of a different organization.

```bash
bal run -- -Corg-name.module-name.port=9090 
java -jar executable.jar -Corg-name.module-name.port=9090 
```

##### Locating Multiple TOML Files

Configurable values can be provided in multiple TOML files using the `BAL_CONFIG_FILES` environment variable.

The file locations can be specified in the environment variable using an OS-specific separator. The precedence order will be determined by the order in which the files are specified in the environment variable. If such an environment variable is not specified, the file located in the current directory with the file name `Config.toml` will be used.

> ##### Providing TOML Content Through the Environment Variable

Reviewer 5: ```suggestion
##### Providing TOML Content Through Environment Variables
```

or

```suggestion
##### Providing TOML Content via Environment Variables
```

##### Time Package

- Introduced the following APIs to support email-typed string conversions:

    - Converts a given UTC to an email string.

        ```ballerina
        import ballerina/time; 
             
        string emailFormattedString = time:utcToEmailString(time:utcNow());
        ```

>     - Converts a given Civil to an email string.

Reviewer 4: ```suggestion
    - Converts a given `time:Civil` to an email string.
```

        ```ballerina
        import ballerina/time; 
       
        time:Civil civil = check time:civilFromString("2021-04-12T23:20:50.520+05:30[Asia/Colombo]");
        string|time:Error emailDateTime = time:civilToEmailString(civil, "GMT");
        ```

>     - Converts a given email string to Civil.

Reviewer 4: ```suggestion
    - Converts a given email string to `time:Civil`.
```

      
#### Improvements

##### I/O Package Updates

- Improved the print APIs to support string templates.

```ballerina
import ballerina/io;
string val = "John";
io:println(`Hello ${val}!!!`);
io:print(`Hello ${val}!!!`);
```

- Changed streaming APIs to be completed from `nil` return. 

##### MySQL Package Updates

> - Changed the previous SSLConfig Record to SecureSocket Record.

Reviewer 4: ```suggestion
- Changed the previous `SSLConfig` Record to `SecureSocket` Record.
```

- Changed the SSLMode value from `SSL_VERIFY_CERT` to `SSL_VERIFY_CA`.

> ##### XML Data Package Updates

Reviewer 3: Check if XmlData is the package name. If it is - Updates to the xmlData Package (or something like that). I'm not sure, so please check.

Reviewer 3: Same with the other package updates.

Author: @kalaiyarasiganeshalingam

Reviewer 6: @praneesha , @dakshitha Let's use the package name `XmlData` as we can't split the package name?

Author: Updated. Thanks a lot!

xml?|Error x = xmldata:fromJson(data);
```

> #### Java.Arrays Package Updates

Reviewer 3: ```suggestion
##### Java.Arrays Package Updates
```

> ##### WebSub Package Updates

Reviewer 3: Another package name. Updates to the `websub` package or Updated `websub` package.

Author: @niveathika @BuddhiWathsala @kalaiyarasiganeshalingam - Shall we rename the `java.arrays` title to `Java.Arrays Package Updates`?

Author: Updated.

    }

```

> ##### Kafka Package Updates

Reviewer 3: Updates to the `kafka` package or Updated the `kafka` package.

Author: Saying "Kafka Package Updates" should be fine because we have the main titles also as follows:

- Language Updates
- Runtime Updates
- Standard Library Updates
- Code to Cloud Updates
- Developer Tools Updates
- Ballerina Packages Updates
- Breaking Changes

 - The `SecureSocket` record is updated.

>  - The `kafka:Producer`, `kafka:Consumer`, and `kafka:Listener` inits are updated.

Reviewer 5: ```suggestion
 - The `init` methods of the `kafka:Producer`, `kafka:Consumer`, and `kafka:Listener` classes are updated.
```

```ballerina
kafka:Producer kafkaProducer = check new(kafka:DEFAULT_URL, config);
kafka:Producer kafkaProducer = check new (bootstrapServers=”localhost:9092”);
kafka:Producer kafkaProducer = check new(”localhost:9092”);
// Same for listener and consumer initialization
```

> ##### NATS Package Updates

Reviewer 3: Updates to the nats package (same as above suggestions).

Author: Resolving as per https://github.com/ballerina-platform/ballerina-dev-website/pull/2093#discussion_r606886986.

>  - The client and listener inits are updated.

Reviewer 5: ```suggestion
 - The `init` methods of the client and listener are updated.
```

```ballerina
nats:Client client = check new(url=”http://google.com:9090”, ssl=config);
nats:Client client = check new(nats:DEFAULT_URL);
nats:Client client = check new(”http://google.com:9090”);
// Same for listener initialization
```

> ##### STAN Package Updates

Reviewer 3: Please edit as mentioned in the previous suggestions about naming packages.

Author: Resolving as per https://github.com/ballerina-platform/ballerina-dev-website/pull/2093#discussion_r606886986.

>  - The client and listener inits are updated.

Reviewer 5: ```suggestion
 - The `init` methods of the client and listener are updated.
```

```ballerina
stan:Client client = check new(url=”http://localhost:9090”);
nats:Client client = check new(nats:DEFAULT_URL);
// Same for listener initialization
```

> ##### RabbitMQ Package Updates

Reviewer 3: Please edit as mentioned in the previous suggestions about naming packages.

Author: Resolving as per https://github.com/ballerina-platform/ballerina-dev-website/pull/2093#discussion_r606886986.

>  - The client and listener inits are updated.

Reviewer 5: ```suggestion
 - The  `init` methods of the client and listener are updated.
```

```ballerina
kafka:Producer kafkaProducer = check new(kafka:DEFAULT_URL, config);
kafka:Producer kafkaProducer = check new (bootstrapServers=”localhost:9092”);
kafka:Producer kafkaProducer = check new(”localhost:9092”);
// Same for listener and consumer initialization
```

##### `nats` Package 

 - The `init` methods of the client and listener are updated.

```ballerina
nats:Client client = check new(url=”http://google.com:9090”, ssl=config);
nats:Client client = check new(nats:DEFAULT_URL);
nats:Client client = check new(”http://google.com:9090”);
// Same for listener initialization
```

##### `stan` Package 

 - The `init` methods of the client and listener are updated.

```ballerina
stan:Client client = check new(url=”http://localhost:9090”);
nats:Client client = check new(nats:DEFAULT_URL);
// Same for listener initialization
```

##### `rabbitmq` Package 

 - The  `init` methods of the client and listener are updated.

```ballerina
rabbitmq:Client client = check new(host=”localhost”, port=9090);
rabbitmq:Client client = check new(rabbitmq:DEFAULT_HOST, rabbitmq:DEFAULT_PORT);
rabbitmq:Client client = check new(”localhost”, 9090);
// Same for listener initialization
```

> 

Reviewer 2: ```suggestion

#### Security Updates
- Removed encrypted passwords and hashed passwords support for Basic Auth file user store authentication.
```

Author: Fixed.

#### Bug Fixes

To view bug fixes, see the [GitHub milestone for Swan Lake Alpha4](https://github.com/ballerina-platform/ballerina-standard-library/issues?q=is%3Aclosed+is%3Aissue+milestone%3A%22Swan+Lake+Alpha4%22+label%3AType%2FBug).

### Developer Tools Updates

#### Bug Fixes

To view bug fixes, see the GitHub milestone for Swan Lake Alpha4 of the repositories below.

- [Language](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FDevTools)
- [OpenAPI](https://github.com/ballerina-platform/ballerina-openapi/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22) 

#### Language Server 

To view bug fixes, see the [GitHub milestone for Swan Lake <VERSION>](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FLanguageServer).

> ### Security Updates
> 

Reviewer 2: ```suggestion
```

Author: Fixed.

> Removed encrypted passwords and hashed passwords support for Basic Auth file user store authentication.

Reviewer 2: ```suggestion
```

Author: Fixed.

```

##### Security Updates

>  - Removed encrypted passwords and hashed passwords support for Basic Auth file user store authentication.

Reviewer 3: This is not clear.

Author: @chanakaudaya

Reviewer 2: There was a support to add encrypted or hashed passwords for Basic Auth file user store authentication. We remove that support with the update done for `configurable` feature.

Author: Resolving since the idea is clear.

#### Bug Fixes

To view bug fixes, see the [GitHub milestone for Swan Lake Alpha4](https://github.com/ballerina-platform/ballerina-standard-library/issues?q=is%3Aclosed+is%3Aissue+milestone%3A%22Swan+Lake+Alpha4%22+label%3AType%2FBug).

> ### Code to Cloud Updates

Reviewer 4: All the links in Bug fixes are empty. So do we need this  section as no improvements or bug fixes? @anuruddhal ?

Reviewer 3: No need if there aren't any bug fixes.

Reviewer 7: We haven't done any bug fixes for this release. Let's remove this.

Author: Removed. Thanks a lot!

```

##### Security Updates

 - Removed encrypted passwords and hashed passwords support for Basic Auth file user store authentication.

#### Bug Fixes

To view bug fixes, see the [GitHub milestone for Swan Lake Alpha4](https://github.com/ballerina-platform/ballerina-standard-library/issues?q=is%3Aclosed+is%3Aissue+milestone%3A%22Swan+Lake+Alpha4%22+label%3AType%2FBug).

### Developer Tools Updates

#### Bug Fixes

To view bug fixes, see the GitHub milestone for Swan Lake Alpha4 of the repositories below.

- [Language](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FDevTools)

> - [Update Tool](https://github.com/ballerina-platform/ballerina-update-tool/issues?q=is%3Aissue+is%3Aclosed+label%3AType%2FBug+project%3Aballerina-platform%2F32)

Reviewer: ```suggestion
```

> - [OpenAPI](https://github.com/ballerina-platform/ballerina-openapi/issues?q=is%3Aissue+is%3Aclosed+label%3AType%2FBug+milestone%3A%22Ballerina+Swan+Lake+-+Alpha%22) 

Reviewer 8: ```suggestion
- [OpenAPI](https://github.com/ballerina-platform/ballerina-openapi/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22) 
```

Author: Fixed.

### Developer Tools Updates

#### Language Server 

> To view bug fixes, see the [GitHub milestone for Swan Lake <VERSION>](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FLanguageServer).

Reviewer 4: Why this is removed? It seems we have one issue as fixed

Author: @nadeeshaan @prabushi - What is the diff between the 2 labels "Team/LanguageServer" and "Team/DevTools"? Do we have to include both or will the DevTools label filter include LanguageServer related issues as well?

Reviewer 9: @praneesha DevTools are excluding the Language Server issues. Hence we need to track both of them separately, due to this reason we have added the `Team` prefix for the labels.

Author: Added it back. Thanks a lot!

- [Language](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22+label%3AType%2FBug+label%3ATeam%2FDevTools)

> - [Update Tool](https://github.com/ballerina-platform/ballerina-update-tool/issues?q=is%3Aissue+is%3Aclosed+label%3AType%2FBug+project%3Aballerina-platform%2F32)

Reviewer 4: No bugs are listed in update tool or open API links given here? Should we remove them @keizer619 @lnash94

Reviewer: We can remove update tool and @lnash94 will update open api link

Reviewer 8: Shall we replace the OpenAPI link with this link  [https://github.com/ballerina-platform/ballerina-openapi/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22](https://github.com/ballerina-platform/ballerina-openapi/issues?q=is%3Aissue+is%3Aclosed+milestone%3A%22Ballerina+Swan+Lake+-+Alpha4%22)

Author: Removed the Update Tool link and updated the OpenAPI link.

- The `error<*>` syntax has been removed.
- Relational expressions are no longer supported with numeric values when the static types of the operands belong to different ordered types.
- The `lang.array:indexOf` and `lang.array:lastIndexOf` lang library functions cannot be used with values that do not belong to `anydata`.

> - An object used as the iterable value in a `foreach` statement, `from` clause, or `join` clause must be a subtype of `object:Iterable`.

Reviewer 4: Why we are having breaking changes section only for language breaking changes? If we are having such a section shouldn't we mention all the breaking changes including stdlibs and tooling as well?

Author: @MaryamZi

Reviewer 5: Ideally, we should. But IINM, most of the time they go under the main content itself for stdlib. Whereas for lang changes we can have them as bullet points.

Author: @MaryamZi - Then, shud this be a sub-topic under the language updates section?

Reviewer 5: Yeah, we can do that too.

Author: Done.
